`Merge` is not distributive

```ts
type Foo = { foo: "bar" } | { bar: "foo" };
type Bar = { foo: "baz" } | { baz: "qux" };

({ foo: "baz" }) satisfies Merge<Foo, Bar>;
```

results in error:

```
error TS1360: Type '{ foo: "baz"; }' does not satisfy the expected type 'Merge<Foo, Bar>'.
  Type '{ foo: "baz"; }' is not assignable to type '{ bar: "foo"; foo: "baz"; } | { bar: "foo"; baz: "qux"; } | { bar: "foo"; foo: "baz"; } | { bar: "foo"; baz: "qux"; } | { bar: "foo"; foo: "baz"; } | { bar: "foo"; baz: "qux"; } | { bar: "foo"; foo: "baz"; } | { bar: "foo"; baz: "qux"; }'.
    Property 'bar' is missing in type '{ foo: "baz"; }' but required in type '{ bar: "foo"; foo: "baz"; }'.

15 ({ foo: "baz" }) satisfies Merge<Foo, Bar>;
```

However, this does work:

```ts
type DistributedMerge<
  A extends object,
  B extends object,
> = A extends object ? (B extends object ? Merge<A, B> : never) : never;

type Foo = { foo: "bar" } | { bar: "foo" };
type Bar = { foo: "baz" } | { baz: "qux" };

({ foo: "baz" }) satisfies DistributiveMerge<Foo, Bar>;
```

--- a/source/merge.d.ts
+++ b/source/merge.d.ts
@@ -3,9 +3,9 @@ import type {PickIndexSignature} from './pick-index-signature.d.ts';
 import type {Simplify} from './simplify.d.ts';
 
 // Merges two objects without worrying about index signatures.
-type SimpleMerge<Destination, Source> = {
+type SimpleMerge<Destination, Source> = Simplify<{
 	[Key in keyof Destination as Key extends keyof Source ? never : Key]: Destination[Key];
-} & Source;
+} & Source>;
 
 /**
 Merge two types into a new type. Keys of the second type overrides keys of the first type.
@@ -45,9 +45,13 @@ Note: If you want a merge type that more accurately reflects the runtime behavio
 @category Object
 */
 export type Merge<Destination, Source> =
-Simplify<
-	SimpleMerge<PickIndexSignature<Destination>, PickIndexSignature<Source>>
-	& SimpleMerge<OmitIndexSignature<Destination>, OmitIndexSignature<Source>>
->;
+	Destination extends unknown // For distributing `Destination`
+		? Source extends unknown // For distributing `Source`
+			? Simplify<
+				SimpleMerge<PickIndexSignature<Destination>, PickIndexSignature<Source>>
+				& SimpleMerge<OmitIndexSignature<Destination>, OmitIndexSignature<Source>>
+			>
+			: never // Should never happen
+		: never; // Should never happen
 
 export {};

--- a/test-d/merge.ts
+++ b/test-d/merge.ts
@@ -165,3 +165,25 @@ expectType<{
 	t4?: number;
 	list: string[];
 }>(baz);
+
+// Unions
+expectType<{foo: number} | {foo: string; bar: boolean}>(
+	{} as Merge<{foo: string}, {foo: number} | {bar: boolean}>,
+);
+expectType<{foo: string} | {bar: boolean; foo: string}>(
+	{} as Merge<{foo: number} | {bar: boolean}, {foo: string}>,
+);
+expectType<{foo: number} | {foo: string; baz: boolean} | {bar: number; foo: number} | {bar: number; baz: boolean}>(
+	{} as Merge<{foo: string} | {bar: number}, {foo: number} | {baz: boolean}>,
+);
+expectType<
+	| {[x: number]: number; foo: number; bar: string}
+	| {[x: number]: number; bar: bigint; foo: number}
+	| {foo: string; bar: boolean; baz: boolean}
+	| {[x: number]: bigint; bar: boolean; baz: boolean}
+>(
+	{} as Merge<
+		{foo: string; bar: string} | {[x: number]: bigint; bar: bigint},
+		{[x: number]: number; foo: number} | {bar: boolean; baz: boolean}
+	>,
+);

